Sprint 11

YanikMax opened the pull request.

Commit:
Свёрстан Launch Screen по макету;
Свёрстан главный экран с лентой изображений.

Commit:
Добавлена градиентная подложка под дату.
Экран профиля сверстан с помощью кода.
Добавлен экран отдельного изображения с возможность шеринга картинки.

Commit:
Исправил констрейнты аватара.
Изменил цвет статус-бара на белый.
Полностью убран код связанный с градиентной подложкой.
Для ImageListCell @IBOutlet сделал приватными.
При импорте UIKit убрал Foundation.
Константу ShowSingleImageSegueIdentifier переименовал (не знаю что на меня нашло).
Извелечение опционала в ImagesListViewController реализовал через guard let.
При верстке кодом использовал предложенное флоу.

Commit:
Для входа в приложение пользователь авторизуется через OAuth2.0;
В проекте создан файл Constants.swift, в нём есть константы AccessKey, SecretKey, RedirectURI и AccessScope;
Свёрстан контроллер WebView;
Реализован KVO — приложение показывает, как обновляется прогресс загрузки веб-страницы;
Проверка необходимости авторизации выделена в отдельный контроллер SplashViewController;
Добавлена структура для декодинга JSON-ответа от POST-запроса авторизации Unsplash — OAuthTokenResponseBody;
Создан новый класс OAuth2Service.

Commit:
Переименовал константы, которые были написаны с заглавной буквы;
Исправил в сториборде работу кнопки назад;
В WebViewViewController организовал код согласно общепринятым нормам;
Убрал force unwrapping и заменил на guard
Добавил trailingAnchor, если будет слишком длинное имя;
Добавил константы из файла Constants;
Исправил ключ AccessKey (в начале был лишний символ @)
Приложение было зарегистрировано, но слетели галочки разрешений (исправил).

Commit:
Устранено состояние гонки в OAuthService.
Устранено состояние гонки c помощью блокировки UI.
Сделан сетевой слой для экрана профиля.
Пока идёт загрузка данных профиля, показывается Splash Screen.
Для отображения аватарки пользователя использована библиотека Kingfisher.
В проект добавлен SwiftKeychainWrapper.

helecta commented on Image Feed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,110 @@
+    }
+    
+    private func updateProgress() {
+        progressView.progress = Float(webView.estimatedProgress)
```
если использовать progressView.setProgress(…), то можно сделать заполнение прогресс бара более плавным.

helecta commented on Image Feed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,110 @@
+        progressView.isHidden = fabs(webView.estimatedProgress - 1.0) <= 0.0001
+    }
+    
+    @IBAction func didRapBackButton(_ sender: Any?) {
```
опечатка, наверное `didTapBackButton`

helecta commented on Image Feed/Constants.swift:
```diff
@@ -0,0 +1,7 @@
+import Foundation
+
+let AccessKey = "CrGnzEgtPBgkxTHXzgkRmFYVbe4e7pOKsxyfiUAzMHM"
```
Такие константы стоит называть со строчной буквы.  Да, в учебнике было так, но все-таки правильно называть со строчной. С заглавной можно называть только типы и протоколы.
Почитать подробнее можно  [тут](https://github.com/RedMadRobot/RMR-swift-style-guide#naming)  или  [тут](https://github.com/tutu-ru/swift-style-guide#%D0%B8%D0%BC%D0%B5%D0%BD%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)

helecta commented on Image Feed/ImagesList/ImagesListViewController.swift:
```diff
@@ -0,0 +1,96 @@
+        let formatter = DateFormatter()
+        formatter.dateStyle = .long
+        formatter.timeStyle = .none
+        return formatter
```
Если так настроить дата форматтер:
```
private lazy var dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "dd MMMM yyyy"
        formatter.locale = Locale(identifier: "ru_RU")
        return formatter
    }()
```
То будет выглядеть как в макете, без  `г.`

helecta commented on Image Feed/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,75 @@
+import Foundation
+
+struct UserResult: Codable {
+    let profile_image: ProfileImage
```
имена функций и переменных принято объявлять в нижнем `camelCase`, а название типов классов, структур и перечислений в верхнем `CamelCase`. `profile_image` - следует переименовать
https://nemecek.be/blog/90/how-to-decode-snake-case-with-codable  - настройки JSONDecoder

helecta commented on Image Feed/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,75 @@
+}
+
+extension ProfileImageService {
+    static let DidChangeNotification = Notification.Name("ProfileImageServiceDidChangeNotification")
```
Такие константы стоит называть со строчной буквы.

helecta commented on Image Feed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,164 @@
+    private let avatarImageView: UIImageView = {
+        let imageView = UIImageView()
+        imageView.image = UIImage(named: "avatar")
+        imageView.translatesAutoresizingMaskIntoConstraints = false
```
аватар должен быть круглым:
```
        imageView.layer.cornerRadius = 35
        imageView.clipsToBounds = true
```

helecta commented on Image Feed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,164 @@
+        label.text = "Hello, world!"
+        label.textColor = .white
+        label.font = UIFont.systemFont(ofSize: 13, weight: .regular)
+        label.translatesAutoresizingMaskIntoConstraints = false
```
Это поле может быть длинным, в этом случае у тебя текст будет обрезаться.
Лучше тут сделать многострочный вариант строки. Для этого нужно сделать так:
```
label.numberOfLines = 0
```
В этом случае еще нужно добавить trailingAnchor для descriptionLabel

helecta commented on Image Feed/ServiceAccessLayer/OAuth2Service.swift:
```diff
@@ -0,0 +1,164 @@
+    
+    private func authTokenRequest(code: String) -> URLRequest {
+        let parameters: [String: Any] = [
+            "client_id": "CrGnzEgtPBgkxTHXzgkRmFYVbe4e7pOKsxyfiUAzMHM",
```
эти константы должны быть вынесены в файл Constants

helecta commented on Image Feed/SplashViewController.swift:
```diff
@@ -0,0 +1,127 @@
+        } else {
+            // Показать экран авторизации
+            let storyboard = UIStoryboard(name: "Main", bundle: .main)
+            let authViewController = storyboard.instantiateViewController(withIdentifier: "AuthViewController") as! AuthViewController
```
Тут force cast: `storyboard.instantiateViewController(withIdentifier: "AuthViewController") as! AuthViewController`, стоит переписать без него
Желательно не использовать force операции, лучше их избегать. Т.к они могут приводить к падению приложения, считается что падение приложения хуже, чем не рабочее состояние.

helecta commented on Image Feed/SplashViewController.swift:
```diff
@@ -0,0 +1,127 @@
+                UIBlockingProgressHUD.dismiss()
+            case .failure:
+                UIBlockingProgressHUD.dismiss()
+                self.showAlert(title: "Что-то пошло не так(", message: "Не удалось войти в систему")
```
Сейчас у тебя такой поток авторизации:
1. Приложение запускается, показан сплеш экран
2. Если пользователь не авторизован, показывается экран авторизации
3. Пользователь авторизовываются во вебвью
4. Вебвью закрывается и показывается сплеш скрин с лоудингом
5. Успевает показаться еще раз экран авторизации перед тем как будет показана лента

Это происходит потому что в SplashScreenViewController, когда в методе `func authViewController(_ vc: AuthViewController, didAuthenticateWithCode code: String)` делаешь dismiss, то повторно вызывается метод viewDidAppear, а т.к токен еще не успели получить к этому моменту, то будет проскакивать экран авторищзации. 

Причем еще будет проблема и с показом алерта, если произошла ошибка. Алерт не успеет показаться, т.к у тебя сразу будет вызван метод `viewDidAppear`, в котором сразу будет показан экран авторизации.

Как вариант решения этих проблем:
1. Добавить в UIBlockingProgressHUD свойство `isShowing`:
```
final class UIBlockingProgressHUD {
    
    static var isShowing: Bool = false
    
    private static var window: UIWindow? {
        guard let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
              let window = windowScene.windows.first
        else {
            return nil
        }
        return window
    }

    static func show() {
        isShowing = true
        window?.isUserInteractionEnabled = false
        ProgressHUD.show()
    }

    static func dismiss() {
        isShowing = false
        window?.isUserInteractionEnabled = true
        ProgressHUD.dismiss()
    }
}
```

2. Во viewDidAppear добавить проверку:
```
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)

        guard UIBlockingProgressHUD.isShowing == false else { return }
        if let token = OAuth2TokenStorage.shared.token {
            fetchProfile(with: token)
            switchToTabBarController()
        } else {
            guard let authViewController = UIStoryboard(name: "Main", bundle: .main).instantiateViewController(withIdentifier: "AuthViewController") as? AuthViewController else {
                assertionFailure("Failed to show Authentication Screen")
                return
            }

            authViewController.delegate = self
            authViewController.modalPresentationStyle = .fullScreen
            present(authViewController, animated: true)
        }
    }
```

3. После того как будет показан алерт с ошибкой, на нажатие кнопки ok нужно добавить показ экрана авторизации

helecta commented on Image Feed/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,72 @@
+        }
+        
+        // создание GET запроса с заголовком Authorization
+        let url = URL(string: "https://unsplash.com/users")!
```
Лучше переписать без force unwrap

helecta replied:
Неправильная ссылка для получения профиля, должна быть https://unsplash.com/me

helecta commented on Image Feed/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,72 @@
+struct ProfileResult: Codable {
+    let username: String
+    let firstName: String
+    let lastName: String
```
lastName и bio необязательные поля и могут отсутствовать. Сейчас у тебя будет происходить ошибка декодинга если их нет.

Commit:
Добавил использование progressView.setProgress для более плавного заполнения прогресс-бара;
Исправлена опечатка в методе didTapBackButton;
Исправлены названия констант в файле констант;
Перенастроил дата форматтер по макету;
Переименовал константу profile_image в файле ProfileImageService.swift в стиле camelCase;
Переименовал константу DidChangeNotification в файле ProfileImageService.swift в стиле camelCase;
Добавил настройки для аватара, чтобы он был круглым;
Добавил настройки для descriptionLabel, чтобы текст был многострочным и не обрезался;
Вынес значения констант из файла OAuth2Service.swift;
Переписал операцию без force unwrap в файле SplashViewController.swift;
Исправил поток авторизации согласно комментарию;
Переписал операцию без force unwrap в файле ProfileService.swift;
Исправил ссылку для получения профиля в файле ProfileService.swift;
Поля lastName и bio сделал опциональным типом;
Пункт 4.2 в сервис ProfileService не добавлена функция fetchProfile, которая делает запрос GET /me с Bearer Token (но данная функция у меня присутствует в коде).

helecta commented on Image Feed/ServiceAccessLayer/OAuth2Service.swift:
```diff
@@ -0,0 +1,164 @@
+            case .success(let responseBody):
+                let authToken = responseBody.accessToken
+                // Сохраняем Bearer Token в User Defaults
+                UserDefaults.standard.set(authToken, forKey: "BearerToken")
```
У тебя метод делает не свою работу. По названию метод отвечает за получение токена. А у тебя получается, что внутри еще происходит и сохранение. Так лучше не делать. В следующих спринтах, когда будешь изучать архитектуру вы с этим разберетесь. 
Лучше вынести это сохранение в SplashViewController. При успешном получении токена, сначала сохранить его, а затем переключать на switchToTabBarController

helecta replied:
И тут надо не напрямую в UserDefaults сохранять, а использовать твой OAuth2TokenStorage

helecta commented on Image Feed/ServiceAccessLayer/OAuth2TokenStorage.swift:
```diff
@@ -0,0 +1,16 @@
+    
+    var token: String? {
+        get {
+            return UserDefaults.standard.string(forKey: tokenKey)
```
Извини, я не заметила при первом ревью. Ты оказывается оставил сохранение в UserDefaults. А по заданию необходимо переписать на SwiftKeychainWrapper

helecta commented on Image Feed/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,80 @@
+        }
+        
+        // создание GET запроса с заголовком Authorization
+        guard let url = URL(string: "https://unsplash.com/me") else {
```
Оказывается должно быть https://api.unsplash.com/me

helecta commented on Image Feed/SplashViewController.swift:
```diff
@@ -0,0 +1,162 @@
+            
+            switch result {
+            case .success:
+                self.switchToTabBarController()
```
По заданию тут необходимо сразу делать загрузку профиля. Причем, пока он не будет загружен, должен быть показан сплеш скрин.

> получение данных профиля в splash-screen, отображаются актуальные данные на момент запуска;

> Пока идёт загрузка данных профиля, показывается Splash Screen:

helecta commented on Image Feed/SplashViewController.swift:
```diff
@@ -0,0 +1,162 @@
+             fetchProfile(token: token)
+             switchToTabBarController()
+         } else {
+             guard let authViewController = UIStoryboard(name: "Main", bundle: .main).instantiateViewController(withIdentifier: "AuthViewController") as? AuthViewController else {
```
можно использовать твой метод showAuthenticationScreen

helecta commented on Image Feed/SplashViewController.swift:
```diff
@@ -0,0 +1,162 @@
+         guard UIBlockingProgressHUD.isShowing == false else { return }
+         if let token = OAuth2TokenStorage().token {
+             fetchProfile(token: token)
+             switchToTabBarController()
```
тут получается нужно убрать эту строчку, чтобы выполнялось условие:

> Пока идёт загрузка данных профиля, показывается Splash Screen:

Т.к у тебя есть переключение на switchToTabBarController внутри fetchProfile

helecta commented on Image Feed/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,80 @@
+    
+    enum CodingKeys: String, CodingKey {
+        case username
+        case firstName
```
Тут неправильные ключи:
```
case firstName = "first_name"
        case lastName = "last_name"
```

helecta commented on Image Feed/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,75 @@
+import Foundation
+
+struct UserResult: Codable {
+    let profileImage: ProfileImage
```
не хватает:
```
enum CodingKeys: String, CodingKey {
        case profileImage = "profile_image"
    }
```

helecta commented on Image Feed/SplashViewController.swift:
```diff
@@ -0,0 +1,162 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        view.backgroundColor = .white
```
Почему фон белый? В макете ypBlack

helecta commented on Image Feed/SplashViewController.swift:
```diff
@@ -0,0 +1,162 @@
+        NSLayoutConstraint.activate([
+            splashImageView.centerXAnchor.constraint(equalTo: view.centerXAnchor),
+            splashImageView.centerYAnchor.constraint(equalTo: view.centerYAnchor),
+            splashImageView.widthAnchor.constraint(equalToConstant: 200),
```
Констрейнты ширина и высота лишние, если их убрать будет размер сплела как в макете

helecta commented on Image Feed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,168 @@
+            return
+        }
+        
+        let profileService = ProfileService()
```
ProfileService - стоит сделать синглтоном. Сохранять в нем полученный профиль, а нее получать его каждый раз. Тогда здесь делаешь проверку есть ли профиль, если есть то показываешь данные из него и только если вдруг нет, то тогда пытаешься получить заново.

helecta commented on Image Feed/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,75 @@
+        }
+        
+        // Создаем GET запрос с заголовком Authorization
+        let urlString = "https://unsplash.com/users/\(username)"
```
https://api.unsplash.com/users/

Commit:
Спасибо Вам, Лилия, за расширенные комментарии.
Сохранение токена перенес в SplashViewController;
Сохранение переписал в SwiftKeychainWrapper;
Изменил путь апи профиля и картинки;
Заменил использование метода на showAuthenticationScreen;
Изменил ключи;
Фон макета изменил на YPBlack;
Констрейнты ширины и высоты убрал;
ProfileService сделано синглтоном.

helecta commented on Image Feed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,170 @@
+    }
+    
+    private func fetchProfileAndUpdateUI() {
+        guard let token = OAuth2TokenStorage().token else {
```
Сначала стоит проверить вдруг профиль уже загружен.
Если профиль уже есть, то сразу показывать его. 
Если его нет, то загружать заново.

helecta commented on Image Feed/SplashViewController.swift:
```diff
@@ -0,0 +1,145 @@
+            switch result {
+            case .success(let authToken):
+                self.oauth2TokenStorage.token = authToken
+                self.switchToTabBarController()
```
Не забудь тут добавить получение профиля!
